src/libfsm/print/llvm.c

```diff
@@ -0,0 +1,824 @@
+/*
+ * Copyright 2008-2024 Katherine Flavel
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <inttypes.h>
+
+#include <print/esc.h>
+
+#include <adt/set.h>
+
+#include <fsm/fsm.h>
+#include <fsm/pred.h>
+#include <fsm/walk.h>
+#include <fsm/print.h>
+#include <fsm/options.h>
+#include <fsm/vm.h>
+
+#include "libfsm/internal.h"
+
+#include "libfsm/vm/vm.h"
+
+#include "ir.h"
+
+#define OPAQUE_POINTERS 1
+
+#ifdef OPAQUE_POINTERS // llvm >= 15
+static const char *ptr_i8   = "ptr";
+static const char *ptr_i32  = "ptr";
+static const char *ptr_void = "ptr";
+#else
+static const char *ptr_i8   = "i8*";
+static const char *ptr_i32  = "i32*";
+static const char *ptr_void = "i8*";
+#endif
```

Comment:
Is it worth-while to have any support for non-opaque pointers?

- It seems like in `src/retest/runner.c` you're unconditionally turning on `-opaque-pointers`.
- I gather it's been a while since LLVM started deprecating the non-opaque pointers, and support for the old style was removed entirely as of the LLVM release a year ago.

I'd be inclined to `unifdef` this, personally.

Reply:
Surprisingly yes! CI here in github runs using clang 14 (!!), and so does the base OS image at work. Those support opaque pointer syntax, but don't have optimisation passes, that comes in clang 15. I'm leaving this here until the air clears.